### PR TITLE
Pipeline Lambda

### PR DESCRIPTION
**Pull request by jamieparkinson**
Lots of commits, but not all that much to review. Things that this PR includes:
- A Lambda for the pipeline app, with associated wiring
- Improved packaging of the pipeline Lambda, using esbuild to compile TS and only its required dependencies
- Deployment scripts for the pipeline Lambda
- IAM permissions for CI agents to deploy the pipeline Lambda

### CODE REVIEW

File: .buildkite/pipeline.yml [hunk 1]
```diff
@@ -1,22 +1,24 @@
+env:
+  LIVE_PIPELINE: 2023-03-24
```
Comment (alexwlchan) [inline .buildkite/pipeline.yml:2]: Ooh didn't know you could have env vars in Buildkite pipelines. 👀

File: infrastructure/api_stack/variables.tf [hunk 2]
```diff
@@ -37,3 +25,11 @@ variable "api_version" {
 variable "external_hostname" {
   type = string
 }
+
+variable "network_config" {
```
Comment (alexwlchan) [inline infrastructure/api_stack/variables.tf:29]: nice

File: infrastructure/ci_iam.tf [hunk 3]
```diff
@@ -0,0 +1,47 @@
+locals {
+  ci_role_name = trimprefix(data.aws_arn.ci_role.resource, "role/")
+}
+
+data "aws_arn" "ci_role" {
+  arn = local.ci_role_arn
+}
+
+resource "aws_iam_role_policy" "ci_s3_publish" {
+  name   = "content-s3-publish"
```
Comment (alexwlchan) [inline infrastructure/ci_iam.tf:10]:
```suggestion
  name   = "ci-s3-publish"
```

or just omit the name entirely, I don't think we usually set it do we?

File: infrastructure/ci_iam.tf [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+}
+
+data "aws_iam_policy_document" "ci_s3_publish" {
+  statement {
+    effect = "Allow"
+    actions = [
+      "s3:Get*",
+      "s3:Put*",
+      "s3:Delete*",
+      "s3:ListBucket"
```
Comment (alexwlchan) [inline infrastructure/ci_iam.tf:22]: Why does CI need anything except PutObject?

File: infrastructure/ci_iam.tf [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    ]
+    resources = [
+      aws_s3_bucket.content_packages.arn,
+      "${aws_s3_bucket.content_packages.arn}/*"
+    ]
+  }
+}
+
+resource "aws_iam_role_policy" "ci_lambda_update" {
+  name   = "content-lambda_update"
```
Comment (alexwlchan) [inline infrastructure/ci_iam.tf:32]:
```suggestion
  name   = "ci-lambda_update"
```

File: infrastructure/pipeline_stack/lambda.tf [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+  policy = data.aws_iam_policy_document.secrets_access.json
+}
+
+resource "aws_iam_role_policy_attachment" "lambda_role_attachment" {
+  role       = module.pipeline_lambda.lambda_role.name
+  policy_arn = aws_iam_policy.secrets_access.arn
+}
+
+resource "aws_security_group" "egress" {
+  name        = "content-pipeline-${var.pipeline_date}-egress"
```
Comment (alexwlchan) [inline infrastructure/pipeline_stack/lambda.tf:65]:
```suggestion
  name        = "content-api-${var.pipeline_date}-egress"
```

**Review by alexwlchan — APPROVED** [review]

File: .buildkite/pipeline.yml [hunk 7]
```diff
@@ -29,14 +31,45 @@ steps:
...
           login: true
-      - docker-compose#v3.5.0:
+      - docker-compose#v4.11.0:
           push:
             - api:756629837203.dkr.ecr.eu-west-1.amazonaws.com/weco/content-api:ref.${BUILDKITE_COMMIT}
             - api:756629837203.dkr.ecr.eu-west-1.amazonaws.com/weco/content-api:latest
 
+  - label: "publish: pipeline"
+    #    branches: "main"
+    branches: "lambda-infra"
```
[the remark this answers is not in the document]
Comment (jamieparkinson) [reply .buildkite/pipeline.yml:41]: Good point, yes

File: .buildkite/pipeline.yml (hunk 1 above)
Comment (jamieparkinson) [reply .buildkite/pipeline.yml:2]: Copied off @paul-butcher in https://github.com/wellcomecollection/concepts-pipeline/blame/main/.buildkite/pipeline.yml!

File: infrastructure/ci_iam.tf (hunk 4 above)
Comment (jamieparkinson) [reply infrastructure/ci_iam.tf:22]: This was from when I was having access policy issues, I will restrict it

**Pull request merged**